packages/server/src/threads/automation.ts

```diff
@@ -330,286 +310,224 @@ class Orchestrator {
+
+          const db = context.getAppDB()
+          try {
+            await db.put(metadata)
+          } catch (err) {
+            logging.logAlertWithInfo(
+              "Failed to write automation metadata",
+              db.name,
+              this.automation._id!,
+              err
+            )
+          }
         }
         return this.executionOutput
       }
     )
   }
 
   private async executeSteps(
-    steps: AutomationStep[],
-    pathIdx?: number
-  ): Promise<void> {
+    ctx: AutomationContext,
+    steps: AutomationStep[]
+  ): Promise<AutomationStepResult[]> {
     return tracer.trace(
       "Orchestrator.executeSteps",
       { resource: "automation" },
       async span => {
         let stepIndex = 0
         const timeout =
           this.job.data.event.timeout || env.AUTOMATION_THREAD_TIMEOUT
+        const stepOutputs: AutomationStepResult[] = []
 
         try {
-          await helpers.withTimeout(
-            timeout,
-            (async () => {
-              while (stepIndex < steps.length) {
-                const step = steps[stepIndex]
-                if (step.stepId === AutomationActionStepId.BRANCH) {
-                  // stepIndex for current step context offset
-                  // pathIdx relating to the full list of steps in the run
-                  await this.executeBranchStep(step, stepIndex + (pathIdx || 0))
-                  stepIndex++
-                } else if (step.stepId === AutomationActionStepId.LOOP) {
-                  stepIndex = await this.executeLoopStep(
-                    step,
-                    steps,
-                    stepIndex,
-                    pathIdx
-                  )
-                } else {
-                  if (!this.stopped) {
-                    await this.executeStep(step)
-                  }
-                  stepIndex++
-                }
+          await helpers.withTimeout(timeout, async () => {
+            while (stepIndex < steps.length) {
+              if (this.stopped) {
+                break
               }
-            })()
-          )
+
+              const step = steps[stepIndex]
+              if (step.stepId === AutomationActionStepId.BRANCH) {
+                const [result, ...childResults] = await this.executeBranchStep(
+                  ctx,
+                  step
+                )
+
+                stepOutputs.push(result)
+                stepOutputs.push(...childResults)
+
+                stepIndex++
+              } else if (step.stepId === AutomationActionStepId.LOOP) {
+                const stepToLoop = steps[stepIndex + 1]
+                const result = await this.executeLoopStep(ctx, step, stepToLoop)
+
+                ctx.steps.push(result.outputs)
+                ctx.stepsById[stepToLoop.id] = result.outputs
+                ctx.stepsByName[stepToLoop.name || stepToLoop.id] =
+                  result.outputs
+
+                stepOutputs.push(result)
+                stepIndex += 2
```

Comment:
this is mildly scary looking - a comment/explanation for the odd increment here would be helpful!